ports/esp8266/network_wlan.c

```diff
@@ -170,6 +170,36 @@ static mp_obj_t esp_disconnect(mp_obj_t self_in) {
 }
 static MP_DEFINE_CONST_FUN_OBJ_1(esp_disconnect_obj, esp_disconnect);
 
+static void build_stations_list_cleanup_callback(void *unused) {
+    (void)unused;
+    wifi_softap_free_station_info();
+}
+
+static mp_obj_t build_stations_list(void) {
+    mp_uint_t count = wifi_softap_get_station_num();
+    if (count == 0) {
+        return mp_obj_new_list(0, NULL);
+    }
+    struct station_info *info = wifi_softap_get_station_info();
+    if (info == NULL) {
+        return mp_obj_new_list(0, NULL);
+    }
+    MP_DEFINE_NLR_JUMP_CALLBACK_FUNCTION_1(ctx, build_stations_list_cleanup_callback, NULL);
+    nlr_push_jump_callback(&ctx.callback, mp_call_function_1_from_nlr_jump_callback);
+    mp_obj_list_t *stations = MP_OBJ_TO_PTR(mp_obj_new_list(count, NULL));
+    struct station_info *current = info;
+    size_t station_index = 0;
+    mp_obj_t entry[2] = { 0 };
+    while (current != NULL && station_index < count) {
```

Comment:
Do you need to check both `current` and `station_index < count`?  If the loop exists early because `current` is `NULL` then the returned list will be short, and entries at the end will be NULL and lead to crashes in the runtime.

You could either not test for `current` being NULL, or use `mp_obj_list_append()`.

If the latter, you could possibly omit the checks for `count == 0` and `info == NULL` above (and not need to create and return an empty list), use `mp_obj_new_list(0, NULL)` to create the initial list, append to it, and then pass `info` through as the argument to the NLR callback and test for `info == NULL` in `build_stations_list_cleanup_callback`.

Reply:
I'd say it's not a concern.  If someone finds that it crashes, we can add a NULL check.

Reply:
OK, I pushed the (hopefully) final changes to the PR.  Thanks for the code suggestions, this looks much better now indeed.